Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,577 @@
+  | liftBind {ι : Type u} (op : f ι) (cont : ι → FreeM f α) : FreeM f α
+
+/-
+Disable simpNF lints for auto-generated constructor lemmas, as they don't follow simp normal
+form patterns. The LHS of these lemmas use `FreeM.pure` which simplifies to `pure` via
+`pure_eq_pure`.
+-/
+attribute [nolint simpNF] FreeM.pure.sizeOf_spec FreeM.pure.injEq
+
+universe u v w
+
+namespace FreeM
+
+/-- Map a function over a `FreeM` monad. -/
+def map {α β : Type w} {F : Type u → Type v} (f : α → β) : FreeM F α → FreeM F β
+  | .pure a => .pure (f a)
+  | .liftBind op cont => .liftBind op (fun z => FreeM.map f (cont z))
+
+instance {F : Type u → Type v} : Functor (FreeM F) where
+  map := .map
+
+instance {F : Type u → Type v} : LawfulFunctor (FreeM F) where
+  map_const := rfl
+  id_map x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp [map, ih]
+  comp_map g h x := by
+    simp [Functor.map]
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih => simp [FreeM.map, ih]
+
+/-- Bind operation for the `FreeM` monad. -/
+protected def bind {a b : Type w} {F : Type u → Type v} (x : FreeM F a) (f : a → FreeM F b) :
+  FreeM F b :=
+  match x with
+  | .pure a => f a
+  | .liftBind op cont => .liftBind op (fun z => FreeM.bind  (cont z) f)
+
+/-- Lift an operation from the effect signature `f` into the `FreeM f` monad. -/
+@[simp]
+def lift {F : Type u → Type v} {ι : Type u} (op : F ι) : FreeM F ι :=
+  .liftBind op FreeM.pure
+
+instance {F : Type u → Type v} : Monad (FreeM F) where
+  pure := .pure
+  bind := .bind
+
+@[simp]
+lemma pure_eq_pure {F : Type u → Type v} {α : Type w} (a : α) :
+    (.pure a : FreeM F α) = pure a := rfl
+
+/-- The `liftBind` constructor is semantically equivalent to `(lift op) >>= cont`. -/
+lemma liftBind_eq_lift_bind {F : Type u → Type v} {ι : Type u} {α : Type u}
+    (op : F ι) (cont : ι → FreeM F α) :
+    .liftBind op cont = (lift op) >>= cont := by rfl
+
+instance {F : Type u → Type v} : LawfulMonad (FreeM F) := LawfulMonad.mk'
+  (bind_pure_comp := by {
+    intro _ _ f x;
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp only [bind, FreeM.bind, Functor.map, Pure.pure, map] at *
+      simp only [ih]
+  })
+  (id_map := id_map)
+  (pure_bind := by intros; rfl)
+  (bind_assoc := by
+    intro _ _ _ x f g
+    induction x with
+    | pure a => rfl
+    | liftBind op cont ih =>
+      simp only [bind, FreeM.bind] at ih ⊢
+      simp [ih])
+
+
+/--
+Interpret a `FreeM f` computation into any monad `m` by providing an interpretation
+function for the effect signature `f`.
+
+This function defines the *canonical interpreter* from the free monad `FreeM f` into the target
+monad `m`. It is the unique monad morphism that extends the effect handler
+`interp : ∀ {β}, F β → M β` via the universal property of `FreeM` -/
+protected def mapM {F : Type u → Type v} {M : Type u → Type w} [Monad M] {α : Type u} :
```

Comment:
This looks more like a `liftM` to me, `mapM` usually takes a `C X` and an `f : X -> m Y` and produces an `m (C Y)`.